### PR TITLE
[MXNET-531] CNN Examples for Scala new API

### PR DESCRIPTION
**Pull request by lanking520**
## Description ##
This PR contains examples written from the new API.
@nswamy @yzhliu @andrewfayres
All examples pushed here are tested locally on my Mac. It lives in the new package called `org.apache.mxnet.examples`
Be aware: CNN Example is specifically designed to run on GPU. Disable the CPU
### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### CODE REVIEW

**Comment by lanking520** [comment]
Currently facing the CI problem as shown below:
```
- Example CI - CNN Example *** FAILED ***

  java.nio.charset.MalformedInputException: Input length = 1

  at java.nio.charset.CoderResult.throwException(CoderResult.java:281)

  at sun.nio.cs.StreamDecoder.implRead(StreamDecoder.java:339)

  at sun.nio.cs.StreamDecoder.read(StreamDecoder.java:178)

  at java.io.InputStreamReader.read(InputStreamReader.java:184)

  at java.io.BufferedReader.read1(BufferedReader.java:210)

  at java.io.BufferedReader.read(BufferedReader.java:286)

  at java.io.Reader.read(Reader.java:140)

  at scala.io.BufferedSource.mkString(BufferedSource.scala:96)

  at org.apache.mxnet.examples.cnntextclassification.DataHelper$.loadMRDataAndLabels(DataHelper.scala:53)

  at org.apache.mxnet.examples.cnntextclassification.DataHelper$.loadMSDataWithWord2vec(DataHelper.scala:166)

  ...
```
I add a fix on codec to see if we can solve that...  Here is the [link](https://stackoverflow.com/questions/13625024/how-to-read-a-text-file-with-mixed-encodings-in-scala-or-java) I found helpful

**Comment by lanking520** [comment]
The previous problem solved, currently facing the memory leak issues:
```
- Example CI - CNN Example *** FAILED ***

  org.apache.mxnet.MXNetError: [23:56:58] src/storage/./pooled_storage_manager.h:118: cudaMalloc failed: out of memory



Stack trace returned 10 entries:

[bt] (0) /work/mxnet/scala-package/native/linux-x86_64-gpu/target/libmxnet-scala-linux-x86_64-gpu.so(dmlc::StackTrace[abi:cxx11]()+0x1bc) [0x7f74407f0f1c]

[bt] (1) /work/mxnet/scala-package/native/linux-x86_64-gpu/target/libmxnet-scala-linux-x86_64-gpu.so(dmlc::LogMessageFatal::~LogMessageFatal()+0x28) [0x7f74407f2138]

[bt] (2) /work/mxnet/scala-package/native/linux-x86_64-gpu/target/libmxnet-scala-linux-x86_64-gpu.so(mxnet::storage::GPUPooledStorageManager::Alloc(mxnet::Storage::Handle*)+0x159) [0x7f744342e7c9]

[bt] (3) /work/mxnet/scala-package/native/linux-x86_64-gpu/target/libmxnet-scala-linux-x86_64-gpu.so(mxnet::StorageImpl::Alloc(mxnet::Storage::Handle*)+0x5d) [0x7f744343084d]

[bt] (4) /work/mxnet/scala-package/native/linux-x86_64-gpu/target/libmxnet-scala-linux-x86_64-gpu.so(mxnet::NDArray::CheckAndAlloc() const+0x238) [0x7f74409a9048]

[bt] (5) /work/mxnet/scala-package/native/linux-x86_64-gpu/target/libmxnet-scala-linux-x86_64-gpu.so(+0x33c19f0) [0x7f7442f429f0]

[bt] (6) /work/mxnet/scala-package/native/linux-x86_64-gpu/target/libmxnet-scala-linux-x86_64-gpu.so(mxnet::imperative::PushFCompute(std::function<void (nnvm::NodeAttrs const&, mxnet::OpContext const&, std::vector<mxnet::TBlob, std::allocator<mxnet::TBlob> > const&, std::vector<mxnet::OpReqType, std::allocator<mxnet::OpReqType> > const&, std::vector<mxnet::TBlob, std::allocator<mxnet::TBlob> > const&)> const&, nnvm::Op const*, nnvm::NodeAttrs const&, mxnet::Context const&, std::vector<mxnet::engine::Var*, std::allocator<mxnet::engine::Var*> > const&, std::vector<mxnet::engine::Var*, std::allocator<mxnet::engine::Var*> > const&, std::vector<mxnet::Resource, std::allocator<mxnet::Resource> > const&, std::vector<mxnet::NDArray*, std::allocator<mxnet::NDArray*> > const&, std::vector<mxnet::NDArray*, std::allocator<mxnet::NDArray*> > const&, std::vector<unsigned int, std::allocator<unsigned int> > const&, std::vector<mxnet::OpReqType, std::allocator<mxnet::OpReqType> > const&)::{lambda(mxnet::RunContext)#1}::operator()(mxnet::RunContext) const+0x1d5) [0x7f7442f6a0c5]

[bt] (7) /work/mxnet/scala-package/native/linux-x86_64-gpu/target/libmxnet-scala-linux-x86_64-gpu.so(+0x3896c7b) [0x7f7443417c7b]

[bt] (8) /work/mxnet/scala-package/native/linux-x86_64-gpu/target/libmxnet-scala-linux-x86_64-gpu.so(mxnet::engine::ThreadedEngine::ExecuteOprBlock(mxnet::RunContext, mxnet::engine::OprBlock*)+0x8e5) [0x7f7443411575]

[bt] (9) /work/mxnet/scala-package/native/linux-x86_64-gpu/target/libmxnet-scala-linux-x86_64-gpu.so(void mxnet::engine::ThreadedEnginePerDevice::GPUWorker<(dmlc::ConcurrentQueueType)0>(mxnet::Context, bool, mxnet::engine::ThreadedEnginePerDevice::ThreadWorkerBlock<(dmlc::ConcurrentQueueType)0>*, std::shared_ptr<dmlc::ManualEvent> const&)+0xeb) [0x7f744342869b]

  at org.apache.mxnet.Base$.checkCall(Base.scala:131)

  at org.apache.mxnet.NDArray.internal(NDArray.scala:942)

  at org.apache.mxnet.NDArray.toArray(NDArray.scala:935)

  at org.apache.mxnet.NDArrayFuncReturn.toArray(NDArray.scala:1152)

  at org.apache.mxnet.examples.cnntextclassification.CNNTextClassification$$anonfun$trainCNN$1$$anonfun$apply$mcVI$sp$1.apply$mcVI$sp(CNNTextClassification.scala:150)

  at scala.collection.immutable.Range.foreach$mVc$sp(Range.scala:160)

  at org.apache.mxnet.examples.cnntextclassification.CNNTextClassification$$anonfun$trainCNN$1.apply$mcVI$sp(CNNTextClassification.scala:127)

  at scala.collection.immutable.Range.foreach$mVc$sp(Range.scala:160)

  at org.apache.mxnet.examples.cnntextclassification.CNNTextClassification$.trainCNN(CNNTextClassification.scala:121)

  at org.apache.mxnet.examples.cnntextclassification.CNNTextClassification$.test(CNNTextClassification.scala:248)

  ...

```
Will try to look into the code to see if we are not disposing some ndarrays

**Comment by lanking520** [comment]
Currently, I have switched from `epoch=200` into 30 and set dropout to 0.0 to match Python configuration. It seemed the training result doesn't looks good

**Comment by nswamy** [comment]
@lanking520 
Though not ideal this example is not tuned to achieve great accuracy. I think we should call this out explicitly in the README that the example is for illustration purposes only at this time.

File: scala-package/examples/src/main/scala/org/apache/mxnet/examples/cnntextclassification/README.md [hunk 1]
```diff
@@ -0,0 +1,23 @@
+# CNN Text Classification Example for Scala
+This is the example using Scala type-safe api doing CNN text classification. 
+Currently, I cannot reproduce the same result in the python example here.
```
Comment (nswamy) [inline scala-package/examples/src/main/scala/org/apache/mxnet/examples/cnntextclassification/README.md]: This example is only for Illustration and not modeled to achieve the best accuracy.

File: scala-package/examples/src/main/scala/org/apache/mxnetexamples/cnntextclassification/CNNTextClassification.scala [hunk 2]
```diff
@@ -17,78 +17,81 @@
...
-      if (dropout > 0f) Symbol.Dropout()()(Map("data" -> hPool, "p" -> dropout))
+      if (dropout > 0f) Symbol.api.Dropout(data = Some(hPool), p = Some(dropout))
       else hPool
     }
 
-    val fc = Symbol.FullyConnected()()(Map("data" -> hDrop, "num_hidden" -> numLabel))
-    val sm = Symbol.SoftmaxOutput()()(Map("data" -> fc, "label" -> inputY))
+    val fc = Symbol.api.FullyConnected(data = Some(hDrop), num_hidden = numLabel)
+    val sm = Symbol.api.SoftmaxOutput(data = Some(fc), label = Some(inputY))
+    fc.dispose()
```
Comment (nswamy) [inline scala-package/examples/src/main/scala/org/apache/mxnetexamples/cnntextclassification/CNNTextClassification.scala]: have you checked that this does not crash. Is `SoftmaxOutput ` using during training

Comment (lanking520) [reply scala-package/examples/src/main/scala/org/apache/mxnetexamples/cnntextclassification/CNNTextClassification.scala]: Yes, I have. Not crashing

**Comment by yzhliu** [comment]
Can we merge?

**Pull request merged**